pkg/controller/certificate-shim/controller.go

```diff
@@ -38,15 +42,71 @@ import (
 )
 
 const (
-	// ControllerName is the name of the ingress-shim controller.
-	ControllerName = "ingress-shim"
+	// IngressShimControllerName is the name of the ingress-shim controller.
+	IngressShimControllerName = "ingress-shim"
+
+	// GatewayShimControllerName is the name of the gateway-shim controller.
+	GatewayShimControllerName = "gateway-shim"
 )
 
 type defaults struct {
 	autoCertificateAnnotations          []string
 	issuerName, issuerKind, issuerGroup string
 }
 
+type ingressShim struct {
+	controller
+}
+
+func (i *ingressShim) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	i.log = logf.FromContext(ctx.RootContext, IngressShimControllerName)
+
+	// create a queue used to queue up items to be processed
+	i.queue = workqueue.NewNamedRateLimitingQueue(controllerpkg.DefaultItemBasedRateLimiter(), IngressShimControllerName)
+
+	ingressInformer := ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses()
+	i.objectInformer = ingressInformer.Informer()
+
+	i.objectLister = &internalIngressLister{ingressInformer.Lister()}
+
+	return i.sharedRegister(ctx)
+}
+
+type gatewayShim struct {
+	controller
+}
+
+func (g *gatewayShim) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	g.log = logf.FromContext(ctx.RootContext, GatewayShimControllerName)
+
+	g.queue = workqueue.NewNamedRateLimitingQueue(controllerpkg.DefaultItemBasedRateLimiter(), GatewayShimControllerName)
+
+	// Discover if the gateway api is available
+	d, err := discovery.NewDiscoveryClientForConfig(ctx.RESTConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s: couldn't construct discovery client: %w", GatewayShimControllerName, err)
+	}
+	resources, err := d.ServerResourcesForGroupVersion(gatewayapi.GroupVersion.String())
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s: couldn't discover gateway API resources: %w", GatewayShimControllerName, err)
+	}
+	if len(resources.APIResources) == 0 {
+		return nil, nil, fmt.Errorf("%s: no gateway API resources were discovered (are the Gateway API CRDS installed?)", GatewayShimControllerName)
+	}
+
+	// 10 hours is default: https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408500629
+	gatewayInformerFactory := gatewayinformers.NewSharedInformerFactory(gatewayclient.NewForConfigOrDie(ctx.RESTConfig), 10*time.Hour)
```

Comment:
question: how are configured the other informers? Are they also set to a 10 hours auto-resync period?

Reply:
Yes, it's in this unexported constant: https://github.com/jetstack/cert-manager/blob/v1.4.0/cmd/controller/app/controller.go#L57

Perhaps I should export it and use it here?